tests/test_data.py

```diff
@@ -196,3 +198,198 @@ def test_zero_length_fails():
         )
+
+
+DICT_DATASET_PARAMS = [
+    (dataset.EpochOrderDictDataset, {"shuffle": False}),
+    (dataset.EpochOrderDictDataset, {"shuffle": True}),
+    (dataset.RandomDictDataset, {}),
+]
+
+DATA_MAP = [
+    {"a": np.zeros([10, 2], dtype=int), "b": np.random.standard_normal([10])},
+    {"asdf": np.zeros([100, 2, 4])},
+    {"asdf": np.zeros([97, 2, 4]), "foo": np.ones([97, 1], dtype=bool)},
+]
+
+
+@pytest.fixture(params=DICT_DATASET_PARAMS)
+def dict_dataset_params(request):
+    return request.param
+
+
+@pytest.fixture(params=DATA_MAP)
+def data_map(request):
+    return copy.deepcopy(request.param)
+
+
+@pytest.fixture
+def dict_dataset(dict_dataset_params, data_map):
+    dataset_cls, kwargs = dict_dataset_params
+    return dataset_cls(data_map, **kwargs)
+
+
+def test_dict_dataset_copy_data_map(dict_dataset: dataset.DictDataset, data_map):
+    """Check that `dict_dataset.data_map` is unchanged by writes to `data_map` param.
+
+    Note that the `data_map` fixture supplies the same instance `data_map` that was used
+    to initialize the `dict_dataset` fixture.
+    """
+    backup_map = copy.deepcopy(dict_dataset.data_map)
+    npt.assert_equal(backup_map, dict_dataset.data_map)
+    for v in data_map.values():
+        # Multiply array in-place by 2.
+        np.multiply(v, v.dtype.type(2), out=v)
+    npt.assert_equal(backup_map, dict_dataset.data_map)
+
+
+def test_dict_dataset_dtypes(dict_dataset, data_map, max_batch_size=80, n_checks=20):
+    """Check that DictDataset preserves array dtype."""
+    for _ in range(n_checks):
+        n_samples = np.random.randint(max_batch_size) + 1
+        sample = dict_dataset.sample(n_samples)
+        for k in data_map.keys():
+            assert data_map[k].dtype == sample[k].dtype
+
+
+@pytest.mark.parametrize("n_samples", [-i for i in range(4)])
+def test_dict_dataset_nonpositive_samples_error(dict_dataset, n_samples):
+    """Check that DictDataset errors on n_samples<=0"""
+    with pytest.raises(ValueError, match="n_samples"):
+        dict_dataset.sample(n_samples)
+
+
+def test_dict_dataset_data_map_error(dict_dataset_params):
+    """Check that DictDataset errors on unequal number of data_map rows."""
+    dataset_cls, kwargs = dict_dataset_params
+    with pytest.raises(ValueError, match="Empty.*"):
+        dataset_cls({}, **kwargs)
+
+    unequal_data_map = {
+        "a": np.zeros([10]),
+        "b": np.zeros([10, 1]),
+        "c": np.zeros([10, 2]),
+        "d": np.zeros([2, 10]),
+    }
+    with pytest.raises(ValueError, match="Unequal.*"):
+        dataset_cls(unequal_data_map, **kwargs)
+
+
+def test_dict_dataset_parallel_rows(
+    dict_dataset_params, max_batch_size=80, n_checks=20
+):
+    """Check that 'parallel rows' among different key-value pairs remain parallel.
+
+    Nontrivially, shuffled datasets should maintain this order.
+    """
+    dataset_cls, kwargs = dict_dataset_params
+    range_data_map = {k: np.arange(50,) for k in "abcd"}
```

Comment:
If I was paranoid, I could worry this wouldn't catch the bug where it just always returns the same value for "a", "b", "c" and "d" keys. Perhaps better if you shifted them slightly -- e.g. i + np.arange(50,) for i, k in enumerate("abcd"). Then checked later they had the expected offset.